bootloader/assets: add fallback entry to grub.cfg

alfonsosanchezbeato opened the pull request:
Add fallback entry to grub configuration, so in updates we boot the
previous kernel if the new one fails to be loaded by grub or the
systemd stub. This is needed when failures happen before exiting UEFI
boot services: otherwise we get stuck in a grub menu and a manual
reboot would be needed. The case when the kernel itself fails should
be handled by a hw watchdog.

Also, remove the test for the kernel file so the fallback mechanism
also works when the try-kernel.efi link has not been
created (previously we would get stuck in grub prompt if that
happened).

pedronis commented on bootloader/assets/data/grub.cfg:
```diff
@@ -15,11 +15,14 @@ if [ -n "$snapd_full_cmdline_args" ]; then
     # a new kernel got installed
     set kernel_status="trying"
     save_env kernel_status
+    # run fallback if we cannot start the kernel
```
s/fallback/fallback menu entry (#1)/  for clarity

pedronis reviewed (commented):
small comment, I haven't looked at the test yet

alfonsosanchezbeato commented:
In some cases I see this error in the tests:

    Please enter the recovery key for disk /dev/disk/by-partuuid/77d...

I think that snap-bootstrap cannot unseal the key anymore because measures change when running the fallback entry (grub->bad kernel.efi->back to grub->fallback kernel). I found a workaround that consists on rebooting instead of trying immediately the fallback kernel, so measurements restart too when reverting the kernel, and locally the test worked. So, fallback entry would be:
```
menuentry "Fallback on failed update" {
    echo "Cannot start new kernel - booting previous one"
    reboot
}
```
@chrisccoulson could you please comment on this? is this hypotheses correct? if it is, is it possible to change snap-bootstrap in some way so we do not need to reboot and can instead jump directly to the fallback kernel?

bboozzoo commented on bootloader/assets/data/grub.cfg:
```diff
@@ -36,14 +39,18 @@ elif [ -n "$kernel_status" ]; then
-    # nothing to boot :-/
-    echo "missing kernel at $prefix/$kernel!"
-fi
+menuentry "Fallback on failed update" {
```
maybe this should have `--id failed-kernel-load-fallback-entry` and then earlier `set fallback=failed-kernel-load-fallback-entry`. I feel that identifying entries by numbers can become tricky in the long run.

alfonsosanchezbeato replied:
I agree, and I actually tried with the id first, but for some reason that didn't work and I had to use the entry number. I've checked and apparently focal's grub should support it nonetheless, so I would give it a another try, maybe I did something wrong.

alfonsosanchezbeato replied:
FTR, I tried again with --id but no luck. Looking at the code, it seems that this is not really possible, at least for focal (it seems that `grub_strtoul()` gets called to get the value for the "fallback" variable):

https://git.launchpad.net/ubuntu/+source/grub2/tree/grub-core/normal/menu.c?h=applied/ubuntu/focal-updates#n337
https://git.launchpad.net/ubuntu/+source/grub2/tree/grub-core/normal/menu.c?h=applied/ubuntu/focal-updates#n171

bboozzoo replied:
Yeah, unfortunately it's still the same in master: https://github.com/rhboot/grub2/blob/971dd6599d838ed21f38f7261c7bdda59f9bbdae/grub-core/normal/menu.c#L336

bboozzoo commented on bootloader/assets/data/grub.cfg:
```diff
@@ -36,14 +39,18 @@ elif [ -n "$kernel_status" ]; then
-fi
+menuentry "Fallback on failed update" {
+    echo "Cannot start new kernel - booting previous one"
+    # clean-up, this is equivalent to reboot with "trying" already set
+    #set kernel_status=""
+    #save_env kernel_status
```
```suggestion
    # kernel status has already been updated, rebooting now will fail the 
    # pending kernel update
```

bboozzoo commented on bootloader/assets/data/grub.cfg:
```diff
@@ -36,14 +39,18 @@ elif [ -n "$kernel_status" ]; then
+    #set kernel_status=""
+    #save_env kernel_status
+
+    reboot
+    #chainloader $prefix/$kernel_fallback snapd_recovery_mode=run $cmdline_args
```
```suggestion
    # reboot to start anew, so that the TPM measurements check out and keys
    # can be unsealed
    reboot
```

alfonsosanchezbeato commented:
@bboozzoo @pedronis thanks for the reviews, I've repushed now. For the moment I think it is probably better to simply reboot in the fail case instead of chainloading the old kernel, otherwise we would need changes in snap-bootstrap that are probably not worth it, as we would just save a bit of boot time in the case there is a kernel rollback.

bboozzoo reviewed (approved):
LGTM

pedronis commented on tests/nested/core/core20-kernel-failover/task.yaml:
```diff
@@ -40,9 +64,20 @@ execute: |
-  tests.nested wait-for reboot "${boot_id}"
+  # wait for a reboot. Note that failure is immediate for kernel.efi
+  # files with invalid PE headers in the FDE case, as snapd is not
+  # able to reseal the encryption key in that case ("cannot compute
```
is reseal here meant to be unseal?

alfonsosanchezbeato replied:
Right, it is "unseal" indeed, changing that now.

pedronis reviewed (approved):
lgtm, one small question

mardy commented on tests/lib/prepare.sh:
```diff
@@ -532,6 +532,43 @@ EOF
+        --empty)
+            printf "" > "$KERNEL_EFI_PATH"
+            ;;
+        --zeros)
```
nitpick: zeroes

alfonsosanchezbeato replied:
Both are correct apparently: https://oneminuteenglish.org/en/zeros-or-zeroes/ 🙂

mardy replied:
Nice, one never stops learning! But my favourite quote from that page is "Zeros is more popular than Zeros but you can choose either." :-)

mardy reviewed (approved):
LGTM, just nitpicks.

alfonsosanchezbeato commented on tests/nested/core/core20-kernel-failover/task.yaml:
```diff
@@ -19,10 +29,24 @@ prepare: |
   mv pc-kernel_*.snap panicking-initramfs-kernel.snap
 
 execute: |
+  set -x
```
Removed

alfonsosanchezbeato commented on tests/nested/core/core20-kernel-failover/task.yaml:
```diff
@@ -44,9 +68,19 @@ execute: |
+  if [ "$PROBLEM_TYPE" = "nolink" ]; then
+      while ! tests.nested exec sudo rm /run/mnt/ubuntu-boot/EFI/ubuntu/try-kernel.efi
+      do sleep 1
+      done
```
Good point, I've changed this to retry now.

alfonsosanchezbeato commented:
@mvo5 @pedronis I have rebased and fixed a couple of spread tests for this branch, I think it is ready now for landing (I just pushed a couple of minor changes so we need to wait for tests though).

mardy reviewed (approved):
Thanks for the updates, LGTM!
